test/src/derive.rs

```diff
@@ -232,6 +251,157 @@ impl IDevId {
     }
 }
 
+pub struct Mek {
+    pub mek: [u8; 64],
+    pub checksum: [u8; 16],
+}
+
+pub struct OcpLockKeyLadderBuilder {
+    cdi: [u32; 16],
+    hek: Option<[u32; 16]>,
+    mdk: Option<[u32; 16]>,
+    intermediate_mek_secret: Option<[u32; 16]>,
+}
+
+impl From<IDevId> for OcpLockKeyLadderBuilder {
+    fn from(value: IDevId) -> Self {
+        Self {
+            cdi: value.cdi,
+            hek: None,
+            mdk: None,
+            intermediate_mek_secret: None,
+        }
+    }
+}
+
+impl OcpLockKeyLadderBuilder {
+    pub fn new(output: DoeOutput) -> Self {
+        let idevid = IDevId::derive(&output);
+        Self {
+            cdi: idevid.cdi,
+            hek: None,
+            mdk: None,
+            intermediate_mek_secret: None,
+        }
+    }
+
+    pub fn add_hek(self, hek_seed: [u32; 8]) -> Self {
+        let mut hek: [u32; 16] = transmute!(hmac512_kdf(
+            swap_word_bytes(&self.cdi).as_bytes(),
+            b"ocp_lock_hek",
+            Some(hek_seed.as_bytes()),
+        ));
+        swap_word_bytes_inplace(&mut hek);
+        Self {
+            hek: Some(hek),
+            ..self
+        }
+    }
+
+    pub fn add_mdk(self) -> Self {
+        let mut mdk: [u32; 16] = transmute!(hmac512_kdf(
+            swap_word_bytes(&self.cdi).as_bytes(),
+            b"ocp_lock_mdk",
+            None
+        ));
+        swap_word_bytes_inplace(&mut mdk);
+
+        Self {
+            mdk: Some(mdk),
+            ..self
+        }
+    }
+
+    pub fn add_intermediate_mek_secret(self, sek: [u8; 32], dpk: [u8; 32]) -> Self {
+        let mut epk: [u32; 16] = transmute!(hmac512_kdf(
+            swap_word_bytes(&self.hek.unwrap()).as_bytes(),
+            b"ocp_lock_epk",
+            Some(sek.as_bytes())
+        ));
+        swap_word_bytes_inplace(&mut epk);
+
+        let mut intermediate_mek_secret: [u32; 16] = transmute!(hmac512_kdf(
+            swap_word_bytes(&epk).as_bytes(),
+            b"ocp_lock_intermediate_mek_secret",
+            Some(dpk.as_bytes()),
+        ));
+        swap_word_bytes_inplace(&mut intermediate_mek_secret);
+
+        Self {
+            intermediate_mek_secret: Some(intermediate_mek_secret),
+            ..self
+        }
+    }
+
+    pub fn derive_mek(self) -> Mek {
+        let ims = self.intermediate_mek_secret.unwrap();
+        let mut mek_secret: [u32; 16] = transmute!(hmac512_kdf(
+            swap_word_bytes(&ims).as_bytes(),
+            b"ocp_lock_derived_mek",
+            None,
+        ));
+        swap_word_bytes_inplace(&mut mek_secret);
+
+        let mek_seed = cmac_kdf(
+            swap_word_bytes(&mek_secret).as_bytes().get(..32).unwrap(),
+            b"ocp_lock_mek_seed",
+            None,
+            4,
+        );
+
+        let key = mek_seed.get(0..32).unwrap().as_bytes();
```

Comment:
nit: If you are going to use `get` and `unwrap` you could also use square brackets. Feel free to update these in a follow up.


```suggestion
        let key = mek_seed[0..32].as_bytes();
```